File: test_programs/timeout.py
#!/usr/bin/env python3
"""
Timeout test program to verify time limits
"""
import time
import signal
import sys

def test_timeouts():
    print("=== Timeout Test ===")
    
    # Test short operation
    print("Testing 2-second operation...")
    start_time = time.time()
    time.sleep(2)
    elapsed = time.time() - start_time
    print(f"✓ 2-second operation completed in {elapsed:.2f}s")
    
    # Test medium operation
    print("Testing 5-second operation...")
    start_time = time.time()
    time.sleep(5)
    elapsed = time.time() - start_time
    print(f"✓ 5-second operation completed in {elapsed:.2f}s")
    
    # Test long operation that should be killed
    print("Testing 30-second operation (should be terminated)...")
    start_time = time.time()
    
    try:
        for i in range(30):
            print(f"Second {i+1}/30")
            time.sleep(1)
        elapsed = time.time() - start_time
        print(f"⚠ 30-second operation completed in {elapsed:.2f}s (not terminated)")
    except KeyboardInterrupt:
        elapsed = time.time() - start_time
        print(f"✓ Operation terminated after {elapsed:.2f}s")

def infinite_loop_test():
    print("=== Infinite Loop Test ===")
    print("Starting infinite loop (should be terminated by timeout)...")
    
    counter = 0
    start_time = time.time()
    
    try:
        while True:
            counter += 1
            if counter % 1000000 == 0:
                elapsed = time.time() - start_time
                print(f"Loop iteration {counter}, elapsed: {elapsed:.2f}s")
    except KeyboardInterrupt:
        elapsed = time.time() - start_time
        print(f"✓ Infinite loop terminated after {elapsed:.2f}s, {counter} iterations")

if __name__ == "__main__":
    if len(sys.argv) > 1 and sys.argv[1] == "infinite":
        infinite_loop_test()
    else:
        test_timeouts()

File: rustbox/src/test_programs/timeout.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <string.h>

void test_timeouts() {
    printf("=== C Timeout Test ===\n");
    
    // Test short operation
    printf("Testing 2-second operation...\n");
    time_t start = time(NULL);
    sleep(2);
    time_t elapsed = time(NULL) - start;
    printf("✓ 2-second operation completed in %ld seconds\n", elapsed);
    
    // Test medium operation
    printf("Testing 5-second operation...\n");
    start = time(NULL);
    sleep(5);
    elapsed = time(NULL) - start;
    printf("✓ 5-second operation completed in %ld seconds\n", elapsed);
    
    // Test long operation that should be killed
    printf("Testing 30-second operation (should be terminated)...\n");
    start = time(NULL);
    
    for (int i = 1; i <= 30; i++) {
        printf("Second %d/30\n", i);
        fflush(stdout);
        sleep(1);
    }
    
    elapsed = time(NULL) - start;
    printf("⚠ 30-second operation completed in %ld seconds (not terminated)\n", elapsed);
}

void infinite_loop_test() {
    printf("=== C Infinite Loop Test ===\n");
    printf("Starting infinite loop (should be terminated by timeout)...\n");
    
    unsigned long long counter = 0;
    time_t start = time(NULL);
    time_t last_print = start;
    
    while (1) {
        counter++;
        
        // Print status every second
        time_t now = time(NULL);
        if (now - last_print >= 1) {
            printf("Loop iteration %llu, elapsed: %ld seconds\n", 
                   counter, now - start);
            fflush(stdout);
            last_print = now;
        }
    }
}

int main(int argc, char *argv[]) {
    if (argc > 1 && strcmp(argv[1], "infinite") == 0) {
        infinite_loop_test();
    } else {
        test_timeouts();
    }
    
    return 0;
}